Задание T1

marsel-mkh opened the pull request:
Все аннотации в стартере. Кеш через Redis. Дока по стартеру в README стартера.
Добавлен jwt.
Добавлена grafana
Добавлен ELK

marsel-mkh committed:
Все первое задание + недопиленное 2

МС-1:
1) Регистрация нового клиента (POST-запрос)
a. Проверка на черные списки
b. Регистрация (создание Client)
c. Создание User (ответ на запрос)
2) Управление продуктами (POST-запрос) (CRUD)
3) Управление продуктами клиентского портфеля (CRUD)
a. DC, CC, NS, PENS
i. Отправка сообщения в топик client_products
b. IPO, PC, AC
i. Отправка сообщения в топик client_credit_products
4) Принимает запрос на создание карты к определенному счету
a. Отправляет запрос с необходимой информацией в топик client_cards
МС-2
1. Слушает топик client_products
a. При получении сообщения создает продукт
2. Слушает топик client_transactions
3. Слушает топик client_cards
a. При получении сообщения создает карту, если счет не заблокирован
МС-3
- Сущности ProductRegistry добавить атрибут monthCount.
1. Слушает топик client_credit_products
a. Условная проверка clientId
i. Обращение к МС-1, GET-запрос за ФИО и номером документа
ii. «Принятие решения»:
1. Если есть существующие кредитные продукты (A, B …), и новый
желаемый продукт в сумме (A + B + C) выходит за лимит N,
устанавливаемый в конфиге, то отказать.
2. Если есть существующие кредитные продукты И новый продукт в
суммарной задолженности по продуктам НЕ выходит за лимит И по
текущим были просрочки, то отказать.
3. Если просрочек в прошлом не было и запрашиваемая сумма
укладывается в лимит N – одобрить.
iii. Отказ или переход к открытию продукта
b. Открытие продукта
c. Создание графика ежемесячных платежей на основе ставки, количества месяцев:

marsel-mkh committed:
Все первое задание + недопиленное 2

МС-1:
1) Регистрация нового клиента (POST-запрос)
a. Проверка на черные списки
b. Регистрация (создание Client)
c. Создание User (ответ на запрос)
2) Управление продуктами (POST-запрос) (CRUD)
3) Управление продуктами клиентского портфеля (CRUD)
a. DC, CC, NS, PENS
i. Отправка сообщения в топик client_products
b. IPO, PC, AC
i. Отправка сообщения в топик client_credit_products
4) Принимает запрос на создание карты к определенному счету
a. Отправляет запрос с необходимой информацией в топик client_cards
МС-2
1. Слушает топик client_products
a. При получении сообщения создает продукт
2. Слушает топик client_transactions
3. Слушает топик client_cards
a. При получении сообщения создает карту, если счет не заблокирован
МС-3
- Сущности ProductRegistry добавить атрибут monthCount.
1. Слушает топик client_credit_products
a. Условная проверка clientId
i. Обращение к МС-1, GET-запрос за ФИО и номером документа
ii. «Принятие решения»:
1. Если есть существующие кредитные продукты (A, B …), и новый
желаемый продукт в сумме (A + B + C) выходит за лимит N,
устанавливаемый в конфиге, то отказать.
2. Если есть существующие кредитные продукты И новый продукт в
суммарной задолженности по продуктам НЕ выходит за лимит И по
текущим были просрочки, то отказать.
3. Если просрочек в прошлом не было и запрашиваемая сумма
укладывается в лимит N – одобрить.
iii. Отказ или переход к открытию продукта
b. Открытие продукта
c. Создание графика ежемесячных платежей на основе ставки, количества месяцев:

marsel-mkh committed:
Задание 3:
МС-2
1. Слушает топик client_products
a. При получении сообщения создает продукт
2. Слушает топик client_transactions
a. Ключ сообщения – UUID
b. По типу транзакции начисляет или списывает сумму транзакции с баланса счета
(если счет не заблокирован и не арестован)
c. Если счет кредитный (isRecalc = true), то создать график платежей Payment на
основе процентной ставки
d. Если сервис получает больше N транзакций по одной карте (cardId) в период
времени Т, то транзакции следует заморозить, счет заблокировать.
e. Если тип транзакции – начисление, счет кредитный, то на основе даты транзакции
И наступил день платежа И на балансе хватает средств, то произвести списание
ежемесячной суммы с баланса.
i. Если суммы недостаточно, то ничего не списывать, платежу проставить
флаг expired (добавить)
3. Слушает топик client_payments
a. Ключ сообщения – UUID
b. Если пришло сообщение, содержащее информацию о платеже И сумма платежа
равна сумме задолженности по кредитному счету, ТО пересчитать баланс счета,
создать платеж Payment, существующие записи Payment по счету оновить,
проставив payed_at = now()

marsel-mkh committed:
Задание 4
1. Разработать annotation based аспект @LogDatasourceError:
a. При возникновении исключения отправляет в топик service_logs
сообщение:
i. В качестве ключа название микросервиса
ii. В заголовок передать пару String type, String value. В качестве
значения использовать ERROR, WARNING, INFO
iii. В сообщении (json-формат) передать: timestamp, сигнатуру
метода, стектрейс исключения, текст исключения, входные
параметры метода, если таковые имеются
b. Если топик недоступен для записи или отправка не получилась, то
аналогичное сообщение записать в таблицу БД error_log
c. Вывести информацию об ошибке в лог консоли
2. Разработать аспект @HttpOutcomeRequestLog (AfterReturnig) для методов,
реализующих отправку http-запросов.
a. В качестве ключа - название микросервиса
b. В заголовке сообщения type = INFO
c. В сообщении timestamp, сигнатура метода, URI запроса, параметры,
тело
3. Разработать Before-аспект @HttpIncomeRequestLog с аналогичной
информацией из п.2
Задание выполнил в отдельном модуле,сделал его стартером и протестил в client-processing модуле.

chasov commented on first-homework/account-processing/src/main/java/com/t1/marselmkh/entity/Account.java:
```diff
@@ -0,0 +1,38 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    private String clientId;
```
Все же лучше указываться явно через @Column

marsel-mkh replied:
Где есть какие либо условия на подобии nullable указывал, если же поле максимально простое и нейминг одинаковый, то смысл? Как будто бойлерплейт

chasov commented on first-homework/account-processing/src/main/java/com/t1/marselmkh/repository/TransactionRepository.java:
```diff
@@ -0,0 +1,12 @@
+
+@Repository
+public interface TransactionRepository extends CrudRepository<Transaction,Long> {
+    int countByCardIdAndTimestampBetween(Long cardId, LocalDateTime from, LocalDateTime now);
```
Лучше использовать обертки

marsel-mkh replied:
методы каунт и экзист всегда же возвращают значение. Как будто обертки в этом случае это избыточно, нет?

marsel-mkh committed:
Задание 5:
Разработать аспект @Metric:

- Измеряет время работы метода

- Если время работы метода выше лимита, установленного в конфиге:

- Отправить сообщение в топик Кафки service_logs с информацией о сигнатуре метода, тип - WARNING, временем работы, входными параметрами, если таковые имеются

- Лимит выполнения устанавливается в application.yml

Разработать аспект @Cached

- Кэширует записи из БД

- Перед выполнением запроса в БД кэш должен проверяться на наличие нужной записи

- Если таковой нет - сделать запрос

- В качестве ключа - primary key, если хранилище индивидуально под сущность, или hashCode объекта, если кэш общий

- Время хранения записи в кэше задать в application.yml

- По истечению времени запись должна удаляться из кэша

Так же чуть рефакторинг

marsel-mkh committed:
1. Реализовать аутентификацию через JWT-токен между сервисами для HTTP-запросов

2. Для операций, направленных на управление продуктами из перечня банковских продуктов, которые может открыть клиент, реализовать ролевую модель. Создание нового продукта может делать пользователь с ролью MASTER, удаление - MASTER, GRAND_EMPLOYEE, редактирование MASTER, GRAND_EMPLOYEE

3. Со стороны клиентского сервиса (МС-1) при создании нового клиента выдавать роль CURRENT_CLIENT

- Если по уже существующему клиенту произошла блокировка (внезапно), то роль менять на BLOCKED_CLIENT и при попытке обратиться к сервису со стороны клиента возвращаться ошибку 403.

marsel-mkh committed:
Merge pull request #2 from marsel-mkh/feature/security

Задание 7 Дедлайн 14.10.2025
1. Реализовать аутентификацию через JWT-токен между сервисами для HTTP-запросов

2. Для операций, направленных на управление продуктами из перечня банковских продуктов, которые может открыть клиент, реализовать ролевую модель. Создание нового продукта может делать пользователь с ролью MASTER, удаление - MASTER, GRAND_EMPLOYEE, редактирование MASTER, GRAND_EMPLOYEE

3. Со стороны клиентского сервиса (МС-1) при создании нового клиента выдавать роль CURRENT_CLIENT

- Если по уже существующему клиенту произошла блокировка (внезапно), то роль менять на BLOCKED_CLIENT и при попытке обратиться к сервису со стороны клиента возвращаться ошибку 403.

marsel-mkh committed:
Задание 8:
1. Покрыть Unit-тестами код, содержащий бизнес-логику в микросервисах

2. Покрыть интеграционными тестами код HTTP-запросов

3. Разработать бизнес-метрики prometheus, вычисляющие количество открытых продуктов в системе (ДК, КК, и пр)

marsel-mkh committed:
Добавлена графана и прометеус, протестирована аннотация

marsel-mkh committed:
прикрутил elk и протестил
